File: nearest_neighbor/nn.java
import java.util.stream.*;
import java.util.*;

public class nn {
    public static void main(String[] args) {
        double[][] points = {{1, 2, 3}, {3, 4, 5}, {5, 6, 7}};
        double[]   qpoint = {3, 6, 5};

        List<Double> vals = new ArrayList<>();
        Arrays.stream(points).forEach(p -> vals.add(IntStream
            .range(0, p.length)
            .mapToObj(i -> Math.pow(p[i] - qpoint[i], 2))
            .reduce(0.0, (Double a, Double b) -> a + b)));
        System.out.println(vals.indexOf(Collections.min(vals)));
    }
}

File: nearest_neighbor/nn.clj
(def points [[1 2 3][3 4 5][5 6 7]])
(def qpoint [3 6 5])
(->> (map vector points (repeat (count points) qpoint))
  (map #(let [[x y] %] (map vector x y)))
  (map (fn [x] (map #(Math/pow (apply - %) 2) x)))
  (apply map vector)(apply map +)(#(.indexOf % (reduce min %)))
  (println)
)

